### PR TITLE
fix: hardware-security returned signatures support for ECDSA256.

### PR DESCRIPTION
**Pull request by merlin-northern**
Changelog: Title
Ticket: MEN-7523

### CODE REVIEW

File: artifact/signer.go [hunk 1]
```diff
@@ -158,16 +160,52 @@ func MarshalECDSASignature(r, s *big.Int) ([]byte, error) {
...
-	s = big.NewInt(0).SetBytes(sig[ecdsa256keySize:])
-	return r, s, nil
+	// in case of a key supplied via PKCS#11 URI, we have no control over what the signature is
+	// since it is designed to be actually verified via the same mechanism (PKCS#11 URI).
+	// We know here that it is ECDSA key, and judging form the size we can assume
+	// that it is ASN.1 encoded. If so, then it should be 72 bytes, the less or equal is here
+	// to support keys of size 71 bytes (strangely seen in the wild). In other words:
+	// if the signature has not been created with MarshalECDSASignature, then we assume
+	// it is to be decoded via ASN.1, with the protection on the signature length.
+	if len(sig) <= ecdsa256Asn1SizeBytes {
```
Comment (lluiscampos) [inline artifact/signer.go:182]: Based on the comment above, and with the aim of giving the best possible error messages to users, this should be:
```suggestion
	if (len(sig) == ecdsa256Asn1SizeBytes) || (len(sig) == ecdsa256Asn1SizeBytes - 1) {
```

File: artifact/signer.go [hunk 2]
```diff
@@ -158,16 +160,52 @@ func MarshalECDSASignature(r, s *big.Int) ([]byte, error) {
...
+			_, err = asn1.Unmarshal(bytesValue, &v)
+			if err != nil {
+				return nil, nil, errors.Wrap(err, "cannot unmarshal asn1 s value")
+			}
+			s = big.NewInt(0).SetBytes(v.Bytes)
+		}
+		return r, s, nil
+	}
+
+	return nil, nil, errors.Errorf("signer: invalid ecdsa key size: %d", len(sig))
```
Comment (lluiscampos) [inline artifact/signer.go:208]: Please consider refactoring the whole thing.
```
func UnmarshalECDSASignature(sig []byte) (r, s *big.Int, e error) {
    if len(sig) == 2*ecdsa256keySize {
        return  UnmarshalECDSASignatureMenderArtifact(...)
    }
    if len(sig) <= ecdsa256Asn1SizeBytes {
        return  UnmarshalECDSASignatureASN1(...)
    }
    return error...
```

**Review by lluiscampos — COMMENTED** [review]
It looks good. The comments are great :rocket:

File: artifact/signer.go (hunk 1 above)
Comment (lluiscampos) [reply artifact/signer.go:182]: Reasoning: a shorter signature should say to the user "invalid ecdsa key size" rather than "cannot unmarshal asn1 data"

File: artifact/signer.go (hunk 2 above)
Comment (lluiscampos) [reply artifact/signer.go:208]: Reasoning: it will make the code more readable and the comments contained inside separate functions

File: artifact/signer.go (hunk 1 above)
Comment (merlin-northern) [reply artifact/signer.go]: > Based on the comment above, and with the aim of giving the best possible error messages to users, this should be:

fixed.

**Comment by merlin-northern** [comment]
@mender-test-bot start pipeline please with cherry on top

**Comment by merlin-northern** [comment]
@lluiscampos heads up: I added smoke test with my public key from my nitro and an artifact signed with the private key from the same nitro :>

File: .gitignore [hunk 3]
```diff
@@ -2,7 +2,6 @@
 *.o
 *.a
 *.so
-*.mender
```
Comment (lluiscampos) [inline .gitignore:5]: Unrelated change

File: ec.pem [hunk 4]
```diff
@@ -0,0 +1,4 @@
+-----BEGIN PUBLIC KEY-----
+MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEDTjFpi9+tJ8KnTPQff34Gpj4E4uc
+T4ML17IzkcWZhtY1O44fqMHmejS0umai8xJlD0Z/THhTiuRO8PC3vpkU9g==
+-----END PUBLIC KEY-----
```
Comment (lluiscampos) [inline ec.pem:4]: This one and `a0-signed-nitro.mender` files should be in a subdirectory inside `tests/`, ideally with an small README on how they have been generated.

File: artifact/signer.go [hunk 5]
```diff
@@ -158,11 +160,56 @@ func MarshalECDSASignature(r, s *big.Int) ([]byte, error) {
...
 
+	// in case of a key supplied via PKCS#11 URI, we have no control over what the signature is
+	// since it is designed to be actually verified via the same mechanism (PKCS#11 URI).
+	// We know here that it is ECDSA key, and judging form the size we can assume
+	// that it is ASN.1 encoded. If so, then it should be 72 bytes, the less or equal is here
+	// to support keys of size 71 bytes (strangely seen in the wild). In other words:
+	// if the signature has not been created with MarshalECDSASignature, then we assume
+	// it is to be decoded via ASN.1, with the protection on the signature length.
+	if len(sig) <= ecdsa256Asn1SizeBytes ||
+		len(sig) <= ecdsa256Asn1SizeBytes-1 {
```
Comment (lluiscampos) [inline artifact/signer.go:179]:
```suggestion
	if len(sig) == ecdsa256Asn1SizeBytes ||
		len(sig) == ecdsa256Asn1SizeBytes-1 {
```

**Review by lluiscampos — CHANGES_REQUESTED** [review]
This fix deserves a better changelog entry :pray: 

For example: "Fixes signature verification with ECDSA keys when the signing has been done externally with PKCS#11. Previously `mender-artifact` would always assume that the signature has been done with the built-in engine, which then wouldn't validate correctly. The bug affected only ECDSA key pairs."

File: artifact/signer.go (hunk 5 above)
Comment (merlin-northern) [reply artifact/signer.go]: wooops sorry, copy-paste.

File: ec.pem (hunk 4 above)
Comment (merlin-northern) [reply ec.pem]: none of the other artifacts in the smoke tests are there. one day we must make order here. what I did last time I was here was to introduce the tests folder and the scripting around the tests. I moved the key and the artifact to the test/data directory and I added a read me.

File: .gitignore (hunk 3 above)
Comment (merlin-northern) [reply .gitignore:5]: without it I will not be able to add the artifacts.

Comment (lluiscampos) [reply .gitignore:5]: Yes, but it deserves a separate commit or a mention in the commit description.

File: ec.pem (hunk 4 above)
Comment (lluiscampos) [reply ec.pem]: > none of the other artifacts in the smoke tests are there. 

What do you mean? Which other Artifacts? As far as I can see this is the only checked in artifact in the repo :mag_right: 

> (...), I moved the key and the artifact to the test/data directory and I added a read me

Thanks

**Review by lluiscampos — APPROVED** [review]

File: ec.pem (hunk 4 above)
Comment (merlin-northern) [reply ec.pem]: artifacts: ignore me it was a complicated situation, you are right. and it would take me so long to properly explain myself. my mistake in general.

File: .gitignore (hunk 3 above)
Comment (merlin-northern) [reply .gitignore:5]: separate commit added

**Comment by merlin-northern** [comment]
I am not running the full pipeline as it already passed with the same code.

**Comment by merlin-northern** [comment]
@mender-test-bot cherry-pick to:
 * 3.11.x

please

**Comment by merlin-northern** [comment]
> Hi 😺 I did my very best, and this is the result of the cherry pick operation:
> 
> * 3.11.x Had merge conflicts, you will have to fix this yourself 😿

https://github.com/mendersoftware/mender-artifact/pull/639

**Comment by merlin-northern** [comment]
> @merlin-northern Is this fix in the version from master that is linked in the documentation? Is there some way I would know when that linked master version is updated?
> 
> https://docs.mender.io/downloads#mender-artifact

you need someone to verify -- both master and the 3.11.x -- I will not issue an official message without double checking. I suppose I can do it, and for sure Client Team can do it. so, take your pick.

**Comment by lluiscampos** [comment]
> @merlin-northern Is this fix in the version from master that is linked in the documentation? Is there some way I would know when that linked master version is updated?
> 
> https://docs.mender.io/downloads#mender-artifact

Yes it is. As well as in the APT repositories experimental distributions (which are updated once a day)
